src/main/java/es/us/isa/restest/generators/ARTestCaseGenerator.java

```diff
@@ -0,0 +1,142 @@
+                    testCases.add(test);
+                }
+                break;
+            default:
+                throw new IllegalArgumentException("The faulty reason '" + faultyReason + "' is not supported.");
+        }
+
+        return test;
+    }
+
+    private TestCase generateARValidTestCase(Operation testOperation) throws RESTestException {
+        Pair<TestCase, Double> bestResult = Pair.with(null, 0.);
+
+        if(testCases.isEmpty()) {
+            bestResult = bestResult.setAt0(generateValidTestCase(testOperation));
+        } else {
+            for(int i = 0; i < getNumberOfCandidates(); i++) {
+                List<TestCase> tcs = new ArrayList<>(testCases);
+                TestCase tc = generateValidTestCase(testOperation);
+                tcs.add(0, tc);
+                Double globalDiversity = getDiversity().evaluate(tcs);
+
+                if(globalDiversity > bestResult.getValue1()) {
+                    bestResult = Pair.with(tc, globalDiversity);
+                }
+            }
+        }
+        return bestResult.getValue0();
+    }
+
+    private TestCase generateARFaultyTestCaseDueToViolatedDependencies(Operation testOperation) throws RESTestException {
+        Pair<TestCase, Double> bestResult = Pair.with(null, 0.);
+
+        if(testCases.isEmpty()) {
+            bestResult = bestResult.setAt0(generateFaultyTestCaseDueToViolatedDependencies(testOperation));
+        } else {
+            for(int i = 0; i < getNumberOfCandidates(); i++) {
+                List<TestCase> tcs = new ArrayList<>(testCases);
+                TestCase tc = generateFaultyTestCaseDueToViolatedDependencies(testOperation);
+                tcs.add(0, tc);
+                Double globalDiversity = getDiversity().evaluate(tcs);
+
+                if(globalDiversity > bestResult.getValue1()) {
+                    bestResult = Pair.with(tc, globalDiversity);
+                }
+            }
+        }
+        return bestResult.getValue0();
+    }
+
+    private TestCase generateARFaultyTestCaseDueToIndividualConstraints(Operation testOperation) throws RESTestException {
+        Pair<TestCase, Double> bestResult = Pair.with(null, 0.);
+
+        if(testCases.isEmpty()) {
+            bestResult = bestResult.setAt0(generateFaultyTestCaseDueToIndividualConstraints(testOperation));
+        } else {
+            for(int i = 0; i < getNumberOfCandidates(); i++) {
+                List<TestCase> tcs = new ArrayList<>(testCases);
+                TestCase tc = generateFaultyTestCaseDueToIndividualConstraints(testOperation);
+                if(tc != null) {
+                    tcs.add(0, tc);
+                    Double globalDiversity = getDiversity().evaluate(tcs);
+
+                    if(globalDiversity > bestResult.getValue1()) {
+                        bestResult = Pair.with(tc, globalDiversity);
+                    }
+                }
+            }
+        }
+        return bestResult.getValue0();
+    }
```

Comment:
These three methods are almost the same, only change lines 68 and 72. Would it be possible to use only one method for the three cases, and an method that returns the specific test case depending on whether it should be valid/invalid?